### DISCUSSION TITLE
Identitystore - throttling limit reached while aggregating user details with describeUser calls

### DISCUSSION THREAD

**Discussion by c1ginly** [aws/aws-sdk-java-v2#3840] [General]
I'm working with the Identitystore service. I have an aggregating endpoint that requires me to fetch a list of all user details along with their groups we have in our Identitystore.
The steps I currently perform are:

1. Fetch the list of all the groups we have
2. Loop through all of these lists, fetch all of their memberships, make a map of User ID - [Group IDs]
3. Loop through the map's entries, perform [describeUser](https://docs.aws.amazon.com/singlesignon/latest/IdentityStoreAPIReference/API_DescribeUser.html) on each User ID

I am testing with 300 groups, and about 2000 users. We run into throttling limits when we are performing the describeUser actions in step 3.
I tried step 3 in parallel, with settings of 10 threads, 4 threads, then finally 2 threads. All of these settings have ended up with us running into throttling errors.

I would like some advices on how to proceed. Is there a better way to get the details of all of my users without performing describeUser calls repeatedly?
If there is no better way. Would I have to implement a retry + backoff strategy? I think there are some predefined strategies built into AWS SDK. I would appreciate some guidance on how to use these predefined strategies with the IdentitystoreClientBuilder.

**Comment by debora-ito** [comment]
Depending on the information you need from describeUser, the same info is available in ListUsers.

Throttling limits are per user and persist for an amount of time, so once you hit the limit retries will not help immediately, unless the backoff time is greater than the ending time of the throttling limit.

I don't have more detailed info on what exactly are the throttling limits for IdentityStore APIs. The service team may have more suggestions, if you can reach out to them.
